### ISSUE TITLE
[question] Deep Filtering in Master-Detail Data Grid using Main Toolbar

### ISSUE THREAD

**Issue by dendycodes** [mui/mui-x#10726]
### The problem in depth 🔍

I'm currently implementing a master-detail data grid in my project. I'd like to request a feature or get guidance on how to deeply filter the data grid using the main toolbar in Material-UI.

### Use Case:
When a user applies a filter on the main data grid (e.g., by Course), the filtering should not only consider the main grid data but also the data from the detailed rows (child grid or expanded details).

### Expected Behavior:

- Users should be able to deeply filter the master-detail data grid using the main toolbar.
- The filter should consider both the main grid data and the data from the detailed rows.
  
### Current Behavior:

- I'm uncertain if there's a direct way to achieve this deep filtering using the current Material-UI data grid components.

I would appreciate any guidance or suggestions on implementing this. If such a feature already exists and I've overlooked it, please guide me to the relevant documentation or examples. Thank you for your time and the excellent library!

<img width="1139" alt="Screenshot 2023-10-19 at 17 26 49" src="https://github.com/mui/mui-x/assets/73158320/6aaecf25-63da-4cc9-ac50-661b8f0526d2">

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Search keywords**: Deep Filtering in Master-Detail Data Grid using Main Toolbar
**Order ID**: 73706

**Comment by cherniavskii** [comment]
Hey @dendycodes 
Interesting use case!

> When a user applies a filter on the main data grid (e.g., by Course), the filtering should not only consider the main grid data but also the data from the detailed rows (child grid or expanded details).

Could you expand a little bit more on this?
What filter criteria would you want to apply to the nested data grid when filtering the root data grid by course?
Are the data for the nested data grid part of the `row` object?

**Comment by dendycodes** [comment]
Hi @cherniavskii

In the main row, we have details pertaining to learners. In the sub-table, I wish to display a list of courses that each learner has either successfully completed or is still in the process of completing. 

The data for the nested data grid is not a part of the main row object. Instead, the master-detail grid represents a completely different table, which is fetched using the main row's learnerID or id.

I aim to filter learners based on the courses.  For example, if a learner is enrolled in "Course A," I would like to be able to filter and display only those learners who are enrolled in "Course A."

**Comment by dendycodes** [comment]
Hi @MBilalShafi,

Currently, the filter options available in the root grid are limited to its own columns. However, I would like to enhance this functionality by including columns from the nested grid within the root grid’s filter options. This way, users will be able to apply filters based on the data present in the columns of the nested grid.

I would greatly appreciate your guidance on how to achieve this integration, ensuring a seamless user experience.

Thank you in advance for your time and support.

**Comment by MBilalShafi** [comment]
In order to understand your use case better, please provide a minimal reproduction test case with the latest version. This would help a lot 👷 .
A live example would be perfect. [This codesandbox.io template](https://codesandbox.io/s/datagrid-v5-quick-start-54iz1) may be a good starting point. Thank you!

**Comment by dendycodes** [comment]
Hi @MBilalShafi  here is the minimal production case  - https://codesandbox.io/s/data-grid-community-forked-wdv522?file=/src/App.tsx

For example, I have two tables: 1) the main table, and 2) the nested table. 
The main table contains three columns: 'id', 'learnerName', and 'age'. The nested table includes two columns: 'id' and 'courseName'.

The feature I need to build will enable users to filter learners by 'courseName' in combination with 'learnerName' and 'age' from the main table.

In other words, how can I add 'className' as a filter option to the main table?

<img width="607" alt="Screenshot 2023-11-06 at 13 21 44" src="https://github.com/mui/mui-x/assets/73158320/8e4b29f5-e00a-4722-8fdf-eed52e586da9">

**Comment by MBilalShafi** [comment]
There isn't a direct way to achieve this since the external grid doesn't recognize the course name column.

To be honest, this seems a more valid use-case of `rowGrouping` rather than master detail. I was wondering if you could use `rowGrouping` here instead of master-detail to achieve the behavior you are trying to achieve, I've tried to update the codesandbox example: https://codesandbox.io/s/data-grid-community-forked-cx3w4k?file=/src/App.tsx

**Comment by jacksonlauder** [comment]
> Hi @dendycodes, thanks for providing the details. I am a bit lost with the scenario tbh.
> 
> Could you explain a bit more about how exactly you want the filters to work? For example, you add a filter on `First name` column from the root grid. Shouldn't that work only on the first name column as it's a column filter?
> 
> Or do you more specifically refer to the [quick filter](https://mui.com/x/react-data-grid/filtering/quick-filter/)? In that case, you may want to use [controlled `filterModel`](https://mui.com/x/react-data-grid/filtering/#controlled-filters) by saving the `filterModel.quickFilterValues` for the root grid in a variable and pass it on to the nested grids.
> 
> Does it make sense?

I have a similar problem to the original reporter and I am trying to implement what you suggested here, could you go into more detail? E.g. How do I set and pass the variable to the nested grids?
